### PR TITLE
Nocloud/OVF seedfrom vendordata

### PR DESCRIPTION
**Pull request by jqueuniet**
This PR modifies the `util.seedfrom` helper function to query optional vendor-data. I started with the idea of implementing for NoCloud, but as OVF is using the same helper, I figured it would be easier to do it for both.

### CODE REVIEW

File: cloudinit/util.py [hunk 1]
```diff
@@ -857,7 +860,17 @@ def read_seeded(base="", ext="", timeout=5, retries=10, file_retries=0):
     if ud_resp.ok():
         ud = ud_resp.contents
 
-    return (md, ud)
+    vd = None
+    try:
+        vd_resp = url_helper.read_file_or_url(vd_url, timeout=timeout,
+                                              retries=retries)
+    except url_helper.UrlError:
```
Comment (lucasmoura) [inline cloudinit/util.py:867]: I think we should log here that there was an url error when reading the vendor-data

File: cloudinit/util.py [hunk 2]
```diff
@@ -857,7 +860,17 @@ def read_seeded(base="", ext="", timeout=5, retries=10, file_retries=0):
...
+        pass
+    else:
+        if vd_resp.ok():
```
Comment (lucasmoura) [inline cloudinit/util.py:870]: same thing here in case we receive a response that is not ok

**Comment by jqueuniet** [comment]
I added some basic logging, but not sure if the messages are good enough, or if I can add something else.

**Review by OddBloke — APPROVED** [review]
LGTM, thanks!

**Comment by jqueuniet** [comment]
This is probably not the right place to talk about this as this PR is long closed, but a functional network is a prerequisite for URL fetching.

This was an issue for us as we use iPXE network boot on diskless bare-metal, and in the end the only feasible option in this scenario is passing the network-config as a kernel cmdline argument.

**Comment by jfgibbins** [comment]
Agreed regarding your place, but I couldn't find where to send an email to bring it up offline. :)  Happy to talk about it elsewhere.  I can be reached at yahoo, same name.  Thanks

**Comment by smoser** [comment]
> This is probably not the right place to talk about this as this PR is long closed, but a functional network is a prerequisite for URL fetching.
> 
> This was an issue for us as we use iPXE network boot on diskless bare-metal, and in the end the only feasible option in this scenario is passing the network-config as a kernel cmdline argument.

well... other datasources can and do get networking information over the network.

They assume "dhcp on first network devcie" and then get the fuller network config.
The same could be done for nocloud.

**Comment by jfgibbins** [comment]
FYI, I'd love to have just put this in as an issue instead of tagging a relevant PR, but that tab seems to be missing on this github project.

@smoser Hey Scott.  Haven't talked with you since working on lxc/lxd.  Currently using vmware ovf to install a seedfrom.  The method of using a cd passed with meta-data, etc is doable, but sloppy, even with bash/govc/cloud-init, and is certainly slower than over the net.  Seedfrom does work decently for grabbing user and vendor, but networking has to resort to a write-files of the network config, as well as a network: config-disabled to remain persistent across reboots.  Sadly VMware really hasn't put any effort into making cloud-init work well through vsphere.  I've looked a lot, but any straight forward methods that don't require create an ISO and attaching has eluded me to date.

Current setup:
govc deploys ova and adds environment variables, including seedfrom (ex, "http://x.x.x.x/*hostname*/")
server boots, grabs user-data and vendor-data
writes previously mentioned files for network config and persists.
meta-data - no use
user-data - host specific config and network-config using write files
vendor-data - generic config
network-config - n/a

Ideal
meta-data - nocloud and ovf don't seem to make any use of usual data found in it (?)
user-data - host specific config
vendor-data - generic config
network-config - obviously network

**Comment by jfgibbins** [comment]
@smoser if I understand what you're saying correctly, its the feature of adding a static route through the dhcp.py(?) for an address such as 169.254.169.254 to reach a web server.  I could certainly see this working.  The things that come to mind are, in an environment like EC2, etc, the local router is actually a software router with additional services, that captures the requests, and by querying IPAM, knows who the client is supposed to be and forwards it to an instance that provides the host specific data.  Unless running on a cloud-stack, or it looks like nsx-t might also provide that service, then the issue is how to know which device it is from vsphere(I include the host name as part of the seedfrom url, with mod_rewrite redirect to a generic user-data if not found) and route to the correct meta/user/vendor instance, and of course, the formatting for network config per the environment.  My current method does work decently, it's just that hang up of write_files for config that seems so kludgy.  My python is extremely limited, as I just started learning it for some basic scripting. (I used python to automate a  remote-trigger blackhole with exaBGP, to update bogons, martians, known c&c networks etc from online sources like team cymru, etc).
I'd love to help where I can to improve the vmware cloud-init anyway I can.  Certainly with lots of testing an my environment(I use a lot of Ubuntu, but a few others such as fedora, etc).  Seems that if ovf could function like nocloud, and nocloud accepted network-config somewhere, the whole thing would be so much better.

**Comment by jqueuniet** [comment]
> FYI, I'd love to have just put this in as an issue instead of tagging a relevant PR, but that tab seems to be missing on this github project.

Issues for cloud-init are tracked on [Launchpad](https://bugs.launchpad.net/cloud-init)

**Pull request merged**